This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [1]:
# get api key from your .env file
import os
from dotenv import load_dotenv, find_dotenv

# Connect the path with your '.env' file name
load_dotenv()

#load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')

Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests
import numpy as np

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [3]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned

url = f'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X?collapse=daily&rows=1&order=asc&start_date=2017-01-01&end_date=2017-12-31&api_key={API_KEY}'
res = requests.get(url)

In [4]:
print(res.text)
#print(res.headers)

{"dataset":{"id":10095370,"dataset_code":"AFX_X","database_code":"FSE","name":"Carl Zeiss Meditec (AFX_X)","description":"Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.\u003cbr\u003e\u003cbr\u003eTrading System: Xetra\u003cbr\u003e\u003cbr\u003eISIN: DE0005313704","refreshed_at":"2020-12-01T14:48:09.907Z","newest_available_date":"2020-12-01","oldest_available_date":"2000-06-07","column_names":["Date","Open","High","Low","Close","Change","Traded Volume","Turnover","Last Price of the Day","Daily Traded Units","Daily Turnover"],"frequency":"daily","type":"Time Series","premium":false,"limit":1,"transform":null,"column_index":null,"start_date":"2017-01-01","end_date":"2017-12-31","data":[["2017-01-02",34.99,35.94,34.99,35.8,null,44700.0,1590561.0,null,null,null]],"collapse":"daily","order":"asc","database_id":6129}}


In [5]:
json = res.json()

In [6]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json

{'dataset': {'id': 10095370,
  'dataset_code': 'AFX_X',
  'database_code': 'FSE',
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'refreshed_at': '2020-12-01T14:48:09.907Z',
  'newest_available_date': '2020-12-01',
  'oldest_available_date': '2000-06-07',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': 1,
  'transform': None,
  'column_index': None,
  'start_date': '2017-01-01',
  'end_date': '2017-12-31',
  'data': [['2017-01-02',
    34.99,
    35.94,
    34.99,
    35.8,
    None,
    44700.0,
    1590561.0,
    None,
    None,
    None]],
  'collapse': 'daily',
  'order': 'asc',
  'database_id': 6129}}

In [7]:
url1 = 'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X?collapse=daily&order=asc&start_date=2017-01-01&end_date=2017-12-31&api_key=qi_mVCvxisRxfza1azEL'
res1 = requests.get(url1)

In [8]:
jsonall = res1.json()

In [16]:
jsonall['dataset']['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [11]:
jsonall['dataset']['data'][0]

['2017-01-02',
 34.99,
 35.94,
 34.99,
 35.8,
 None,
 44700.0,
 1590561.0,
 None,
 None,
 None]

In [12]:
highs = []
lows = []
x=0

for value in jsonall['dataset']['data'] :
    highs.append(jsonall['dataset']['data'][x][2])
    lows.append(jsonall['dataset']['data'][x][3])
    x +=1

maxhigh = max(highs)
minlow = min(lows)
print(f"The highest opening price for AFX_X in 2017 was {maxhigh}.")
print(f"The lowest opening price for AFX_X in 2017 was {minlow}.")

The highest opening price for AFX_X in 2017 was 53.54.
The lowest opening price for AFX_X in 2017 was 33.62.


In [13]:
difference = []
x = 0
for value in jsonall['dataset']['data'] :
    difference.append(jsonall['dataset']['data'][x][2] - jsonall['dataset']['data'][x][3])
    x +=1
    
bigdiff = max(difference)
print(f"The greatest change of price for AFX_X in 2017 in any one day was {bigdiff:.2f}.")

The greatest change of price for AFX_X in 2017 in any one day was 2.81.


In [14]:
closing = []
x = 0

for value in jsonall['dataset']['data'] :
    closing.append(jsonall['dataset']['data'][x][4])
    x +=1
    
anydaydiff = max(closing) - min(closing)
print(f"The greatest change of price for AFX_X in 2017 between any two days regardless of order was {anydaydiff:.2f}.")

y=0
consec_close = []

for num in range(len(closing) - 1) :
    consec_close.append(closing[y+1] - closing[y])
    y +=1

print(f"The greatest change of price for AFX_X in 2017 between any two consecutive days was {max(consec_close):.2f}.")

The greatest change of price for AFX_X in 2017 between any two days regardless of order was 19.03.
The greatest change of price for AFX_X in 2017 between any two consecutive days was 1.72.


In [15]:
trade = []
x = 0

for value in jsonall['dataset']['data'] :
    trade.append(jsonall['dataset']['data'][x][6])
    x += 1

print(f"The average trade volume for AFX_X in 2017 was {np.mean(trade):.2f}.")
print(f"The median trade volume for AFX_X in 2017 was {np.median(trade):.2f}.")

The average trade volume for AFX_X in 2017 was 89124.34.
The median trade volume for AFX_X in 2017 was 76286.00.


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

The highest opening price for AFX_X in 2017 was 53.54.
The lowest opening price for AFX_X in 2017 was 33.62.
The greatest change of price for AFX_X in 2017 in any one day was 2.81.
The greatest change of price for AFX_X in 2017 between any two days regardless of order was 19.03.
The greatest change of price for AFX_X in 2017 between any two consecutive days was 1.72.
The average trade volume for AFX_X in 2017 was 89124.34.
The median trade volume for AFX_X in 2017 was 76286.00.